tech_reports/LLMs/llms.md

```diff
@@ -56,9 +56,123 @@ Authors:
+- Prefill also computes the first token for the following autoregressive generation.
+
+#### Decode
+- The decode phase is parallel for all users, but sequential for each token within a batch of users.
+- Each new token can only be generated after the previous one, as the system must maintain causality in attention computations.
+
+#### Module and Test Differences
+The attention module class (`TtLlamaAttention`) implements two primary methods for attention computation: `forward_prefill` and `forward_decode`. Attention has two different test files, `test_attention_decode` and `test_attention_prefill`, which create the appropriate input tensors:
+- A tensor of size `(batch, dim)` in L1 for decode
+- A tensor of size `(seqlen, dim)` in DRAM for prefill
+
+Each test compares the attention module output and KV-cache correlation between the PyTorch host implementation and the ttnn device implementation.
+
+The MLP module class (`TtLlamaMLP`) handles prefill and decode in the same file but has some technical differences (mentioned below).
+
+The decoder module and model module also handle prefill and decode in the same file, but they call the respective modes within the attention and MLP modules.
+
+#### Technical Implementation Differences
+
+##### 1. Reshaping for Large Matrix Multiplications
+In prefill mode, the system reshapes input tensors to process sequences in smaller chunks in parallel for larger matrix multiplications, such as `wqkv`, `wo` in the attention module, and `w1`, `w2`, `w3` in the MLP module. This reshaping prevents running out of memory in cases of long prefill sequence lengths. For instance:
+
+```python
+if seq_len > 2048:
+    x_11SH = ttnn.reshape(x_11SH, [1, seq_len // 2048, 2048, -1])
+
+xqkv_fused = ttnn.linear(
+    x_11SH,
+    self.wqkv,
+    dtype=ttnn.bfloat16,
+    memory_config=ttnn.DRAM_MEMORY_CONFIG,
+    compute_kernel_config=self.compute_kernel_config_hifi2,
+    program_config=self.model_config["XQKV_PREFILL_PROGCFG"](seq_len),
+)
+```
+
+This reshaping is not needed for decode mode because it processes one token at a time. In decode mode, the parallelization is over user batches, which is only up to 32.
+
+##### 2. KV Cache Management
+###### Prefill:
+```python
+# Fill cache with initial states
+ttnn.experimental.paged_fill_cache(
+    keys_BKSD,
+    k_fill,
+    page_table,
+    batch_idx=user_id
+)
+```
+
+###### Decode:
+```python
+# Update KV cache with a single new token
+ttnn.experimental.paged_update_cache(
+    keys,
+    k_heads_1BKD,
+    update_idxs_tensor=current_pos,
+    page_table=page_table
+)
+```
+
+##### 3. Attention Computation
+###### Prefill:
+```python
+# Split q_heads into num_groups and kv_heads for parallel group computation for GQA
+q_heads_84SD_8b = ttnn.reshape(
+    q_heads_1QSD_8b,
+    [self.n_local_kv_heads, self.n_local_heads // self.n_local_kv_heads, -1, self.head_dim]
```

Comment:
This should be a reshape -> transpose, or even  `nlp_create_heads`

Reply:
why should it be a transpose?

Reply:
```
Q: 1, 1, S, D
Q = Q.reshape(1, S, NH, DH)
Q = Q.permute(0, 2, 1, 3) # 1, NH, S, DH
```

I might be confused because you mention `num_groups` and `kv_heads` and I'm not as familiar with how we've done GQA in the past, but SDPA should support input shapes of 
```
Q: 1, NQH, S, DH
K: 1, NKVH, S, DH
V: 1, NKVH, S, DH
```

Reply:
For GQA prefill, we put kv_heads on dim 0 for k, q, v heads such that, 
* for k_heads and v_heads, NKVH=1 and  B=k_heads
* for q_heads, NQH=num_q_heads/num_k_heads=group size and B=k_heads. 

This way we can do B sdpa MQAs (with q_heads=group and kv_heads=1) in parallel